dhcpv4/bsdp.go

```diff
@@ -0,0 +1,328 @@
+		if opt.Code == BSDPOptionBootImageList {
+			images, err := parseBootImagesFromBSDPOption(opt.Data)
+			if err != nil {
+				return nil, err
+			}
+			bootImages = append(bootImages, images...)
+		}
+	}
+
+	return bootImages, nil
+}
+
+// NewInformListForInterface creates a new INFORM packet for interface ifname
+// with configuration options specified by config.
+func NewInformListForInterface(iface string, replyPort uint16) (*DHCPv4, error) {
+	d, err := NewInformForInterface(iface /* needsBroadcast */, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// These are vendor-specific options used to pass along BSDP information.
+	vendorOpts := []Option{
+		Option{
+			Code: BSDPOptionMessageType,
+			Data: []byte{BSDPMessageTypeList},
+		},
+		Option{
+			Code: BSDPOptionVersion,
+			Data: BSDPVersion1_1,
+		},
+	}
+
+	// If specified, replyPort MUST be a priviledged port.
+	if replyPort != 0 && replyPort != ClientPort {
+		if replyPort >= 1024 {
+			return nil, fmt.Errorf("replyPort must be a priviledged port (< 1024)")
+		}
+		bytes := make([]byte, 3)
+		bytes[0] = 2
+		binary.BigEndian.PutUint16(bytes[1:], replyPort)
+		d.AddOption(Option{
+			Code: BSDPOptionReplyPort,
+			Data: bytes,
+		})
+	}
+	d.AddOption(Option{
+		Code: OptionVendorSpecificInformation,
+		Data: OptionsToBytes(vendorOpts),
+	})
+
+	d.AddOption(Option{
+		Code: OptionParameterRequestList,
+		Data: []byte{OptionVendorSpecificInformation, OptionClassIdentifier},
+	})
+
+	u16 := make([]byte, 2)
+	binary.BigEndian.PutUint16(u16, 1500)
+	d.AddOption(Option{
+		Code: OptionMaximumDHCPMessageSize,
+		Data: u16,
+	})
+
+	vendorClassID, err := makeVendorClassIdentifier()
+	if err != nil {
+		return nil, err
+	}
+	d.AddOption(Option{
+		Code: OptionClassIdentifier,
+		Data: []byte(vendorClassID),
+	})
+
+	d.AddOption(Option{Code: OptionEnd})
+	return d, nil
+}
+
+// InformSelectForAck constructs an INFORM[SELECT] packet given an ACK to the
+// previously-sent INFORM[LIST] with BSDPConfig config.
+func InformSelectForAck(ack DHCPv4, replyPort uint16, selectedImage BootImage) (*DHCPv4, error) {
+	d, err := New()
+	if err != nil {
+		return nil, err
+	}
+	d.SetOpcode(OpcodeBootRequest)
+	d.SetHwType(ack.HwType())
+	d.SetHwAddrLen(ack.HwAddrLen())
+	clientHwAddr := ack.ClientHwAddr()
+	d.SetClientHwAddr(clientHwAddr[:])
+	d.SetTransactionID(ack.TransactionID())
+	if ack.IsBroadcast() {
+		d.SetBroadcast()
+	} else {
+		d.SetUnicast()
+	}
+
+	// Data for BSDPOptionSelectedBootImageID
+	vendorOpts := []Option{
+		Option{
+			Code: BSDPOptionMessageType,
+			Data: []byte{BSDPMessageTypeSelect},
+		},
+		Option{
+			Code: BSDPOptionVersion,
+			Data: BSDPVersion1_1,
+		},
+		Option{
+			Code: BSDPOptionSelectedBootImageID,
+			Data: append([]byte{4}, selectedImage.ID.toBytes()...),
+		},
+	}
+
+	// Find server IP address
+	var serverIP []byte
+	for _, opt := range ack.options {
+		if opt.Code == OptionServerIdentifier {
+			serverIP = make([]byte, 4)
+			copy(serverIP, opt.Data)
```

Comment:
if I recall correctly this does not behave as expected, since `net.IP` is a slice (used for both IPv4 and IPv6) and the last four bytes are used for IPv4, not the first four. This will probably yield a `0.0.0.0` address, see https://play.golang.org/p/ofkyBXT3tTG.

You can use a simpler syntax, just `serverIP = net.IP(opt.Data)`